calico_containers/adapter/diags.py

```diff
@@ -0,0 +1,114 @@
+import os
+from datetime import datetime
+import sh
+import tarfile
+import socket
+import tempfile
+import subprocess
+from datastore import DatastoreClient
+from etcd import EtcdException
+from shutil import copytree
+
+def save_diags(upload=False):
+    # Create temp directory
+    temp_dir = tempfile.mkdtemp()
+    temp_diags_dir = os.path.join(temp_dir, 'diagnostics')
+    os.mkdir(temp_diags_dir)
+    print("Using temp dir: %s" % temp_dir)
+
+    # Write date to file
+    with open(os.path.join(temp_diags_dir, 'date'), 'w') as f:
+        f.write("DATE=%s" % datetime.strftime(datetime.today(),"%Y-%m-%d_%H-%M-%S"))
+
+    # Write hostname to file
+    with open(os.path.join(temp_diags_dir, 'hostname'), 'w') as f:
+        f.write("%s" % socket.gethostname())
+
+    # Write netstat output to file
+    with open(os.path.join(temp_diags_dir, 'netstat'), 'w') as f:
+        try:
+            print("Dumping netstat output")
+            netstat = sh.Command._create("netstat")
+            f.writelines(netstat("-an"))
```

Comment:
When I use flags in code, I like to separate them;
`netstat("-a", "-n")`
or even better, use their expansions;
`netstat("--all", "--numeric")`
or even clearer, comment on what they mean;
```python
netstat(
    # Show both listening and non-listening sockets.
    "--all",
    # Show numerical addresses instead of trying to determine symbolic host, port or user names.
    "--numeric",
)
```
And for maximally idiomatic use of `sh`,
```python
netstat(
    # Show both listening and non-listening sockets.
    all=True,
    # Show numerical addresses instead of trying to determine symbolic host, port or user names.
    numeric=True,
)
```
But of course, this is all in the name of clarity, so use the one you think is clearest. For example I would never expand `ls -a` or even comment on it.